hw/xfree86/common/xf86Xinput.c

```diff
@@ -529,82 +529,121 @@ HostOS(void)
+                    pattern->mode = MATCH_IS_STRSTR;
+            return (strcmp(attr, pattern->str)) ? 0 : -1;
 #endif
 
+        case MATCH_IS_PATHNAME:
 #ifdef HAVE_FNMATCH_H
-static int
-match_path_pattern(const char *attr, const char *pattern)
-{
-    return fnmatch(pattern, attr, FNM_PATHNAME);
-}
+            return (fnmatch(pattern->str, attr, FNM_PATHNAME)) ? 0 : -1;
 #else
-#define match_path_pattern match_substring
+            LogMessageVerb(X_WARNING, 1, "Pathname matching requested but unavailable for \"%s\", resorting to search for substring\n", pattern->str);
+                    pattern->mode = MATCH_IS_STRSTR;
+            return (strcmp(attr, pattern->str)) ? 0 : -1;
+#endif
+        case MATCH_IS_SEQSTR:
+            {
+                char* str = pattern->str;
+                while (*str) {
+                    attr=strstr(attr, str);
+                    if (!attr) return 0;
+                    attr += strlen(str);
+                    str += strlen(str);
+                    str++;
+                }
+            }
+            return -1;
+        case MATCH_IS_REGEX:
+        default:
+#ifdef HAVE_REGEX_H
+            if (pattern->regex == NULL) {
+                int r;
+                if ((pattern->regex = malloc(sizeof(regex_t))) == NULL) {
+                    pattern->mode = MATCH_IS_INVALID;
+                    return 0;
+                }
+                r = regcomp(pattern->regex, pattern->str, REG_EXTENDED | REG_NOSUB);
+                if (r) { /* Wrong regex */
+                    regfree(pattern->regex);
+                    free(pattern->regex);
+                    LogMessageVerb(X_ERROR, 1, "Wrong regex: \"%s\"\n", pattern->str);
+                    pattern->mode = MATCH_IS_INVALID;
+                    return 0;
+                }
+            }
+            return (regexec(pattern->regex, attr,0, NULL, 0)) ? 0 : -1;
+#else
+            LogMessageVerb(X_ERROR, 1, "Regex is unavailable at this system: \"%s\"\n", pattern->str);
+            pattern->mode = MATCH_IS_INVALID;
+            return 0;
 #endif
-
-/*
- * If no Layout section is found, xf86ServerLayout.id becomes "(implicit)"
- * It is convenient that "" in patterns means "no explicit layout"
- */
-static int
-match_string_implicit(const char *attr, const char *pattern)
-{
-    if (strlen(pattern)) {
-        return strcmp(attr, pattern);
-    }
-    else {
-        return strcmp(attr, "(implicit)");
     }
 }
 
 /*
- * Match an attribute against a list of NULL terminated arrays of patterns.
- * If a pattern in each list entry is matched, return TRUE.
+ * Match an attribute against a list of xf86MatchGroup's.
+ * Return TRUE only if each list entry is successful.
  */
-static Bool
-MatchAttrToken(const char *attr, struct xorg_list *patterns,
-               int (*compare) (const char *attr, const char *pattern))
```

Comment:
This previous idea of using `compare` function pointers looks rather nice. Can we keep it somehow? I mean instead of an explicit switch over xf86MatchMode we could just have several "match" functions and pass the function pointers around. Unless there was a serious reason to not do it.

Reply:
I doubt that extra far calls are more efficient than compiler-optimized switch/case statement. 
There're various techniques (depending on length and actual bit values), eg. a tiny map, a mask search tree, etc.
We really had to measure both options to give a definitive answer. But I doubt it's worth that effort on such a cold path.

The whole issue would be different if those matchers were more generic and used in various other places - then we could put them into some more common place. If you actually find good case for that, feel free to submit PR for that ;-)